src/verify.rs

```diff
@@ -28,35 +31,96 @@ impl fmt::Display for VerificationError {
+        Ok(Pubkey::G1(g1_from_fixed_unchecked(data)?))
+    }
+
+    // Creates a public key from a 96 byte representation (point on G2)
+    pub fn from_96(data: [u8; 96]) -> Result<Self, InvalidPoint> {
+        Ok(Pubkey::G2(g2_from_fixed(data)?))
+    }
+
+    /// Creates a public key from a 96 byte representation (point on G2).
+    /// Like [`Pubkey::from_96`] without guaranteeing that the encoding represents a valid element.
+    /// Only use this when you know for sure the encoding is correct.
+    pub fn from_96_unchecked(data: [u8; 96]) -> Result<Self, InvalidPoint> {
+        Ok(Pubkey::G2(g2_from_fixed_unchecked(data)?))
+    }
+
+    /// Creates a public key from a 48 or 96 byte representation (point on G1 or G2).
+    pub fn from_variable(data: &[u8]) -> Result<Self, InvalidPoint> {
+        match data.len() {
+            48 => {
+                let mut buf = [0u8; 48];
+                buf[..].clone_from_slice(data);
+                Pubkey::from_48(buf)
+            }
+            96 => {
+                let mut buf = [0u8; 96];
+                buf[..].clone_from_slice(data);
+                Pubkey::from_96(buf)
+            }
+            other => Err(InvalidPoint::InvalidLength { actual: other }),
+        }
+    }
+}
+
 /// Checks beacon components to see if they are valid.
 ///
 /// For unchained mode set `previous_signature` to an empty value.
 pub fn verify(
-    pk: &G1Affine,
+    pk: &Pubkey,
     round: u64,
     previous_signature: &[u8],
     signature: &[u8],
 ) -> Result<bool, VerificationError> {
-    let msg_on_g2 = verify_step1(round, previous_signature);
-    verify_step2(pk, signature, &msg_on_g2)
+    match pk {
+        Pubkey::G1(pk) => {
+            let msg = message(round, previous_signature);
+            let msg_on_g2 = msg_to_curve_g2(&msg);
+            verify_step2_g2(pk, signature, &msg_on_g2)
+        }
+        Pubkey::G2(pk) => {
+            let msg = message(round, previous_signature);
+            let msg_on_g1 = msg_to_curve_g1(&msg);
+            verify_step2_g1(pk, signature, &msg_on_g1)
+        }
+    }
 }
 
-/// First step of the verification.
-/// Should not be used directly in most cases. Use [`verify`] instead.
-///
-/// This API is not stable.
-#[doc(hidden)]
-pub fn verify_step1(round: u64, previous_signature: &[u8]) -> G2Affine {
-    let msg = message(round, previous_signature);
-    msg_to_curve(&msg)
+fn verify_step2_g1(
+    pk: &G2Affine,
+    signature: &[u8],
+    msg_on_g1: &G1Affine,
+) -> Result<bool, VerificationError> {
+    let g2 = G2Affine::generator();
+    let sigma = match g1_from_variable(signature) {
+        Ok(sigma) => sigma,
+        Err(err) => {
+            return Err(VerificationError::InvalidPoint {
+                field: "signature".into(),
+                msg: err.to_string(),
+            })
+        }
+    };
+    Ok(fast_pairing_equality(&sigma, &g2, msg_on_g1, pk))
```

Comment:
This code is developed trial and error style using the type system. Is it correct?